lightning/src/ln/channelmanager.rs

```diff
@@ -3477,6 +3477,25 @@ where
 		}
 	}
 
+	/// This is a variant of [`ChannelManager::fail_htlc_backwards`] that allows you to specify the
+	/// reason for the failure.
+	/// 
+	/// See https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages for
+	/// valid failure codes and data.
+	pub fn fail_htlc_backwards_with_reason(&self, payment_hash: &PaymentHash, failure_code: u16, failure_data: Vec<u8>) {
```

Comment:
After looking into the [failure messages in BOLT 4](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#failure-messages) I was thinking that there might be cases where the user would want to fail back with an error message besides `incorrect_or_unknown_payment_details`, mainly `final_incorrect_cltv_expiry`, `final_incorrect_htlc_amount`, or a channel update as it mentions in the section about what a final node might error with (not sure if cases where these errors would occur are already handled somewhere else). My implementation doesn't handle these specific errors as of now; I was actually going to ask if that made sense. I was honestly not super sure about the specific situation where this might be needed, and was wondering if you had something in mind when opening #1570?

For passing in an arbitrary code and data, it seems the only way to create an `HTLCFailReason::Reason` at the moment is using an arbitrary code and data, at least from what I can tell [here](https://github.com/lightningdevkit/rust-lightning/blob/main/lightning/src/ln/onion_utils.rs#L643). These seem like they would be good to represent with enums, and then maybe expose that to the user?

Reply:
Sounds good. I'll try to get an implementation of this working sometime in the next 2 days.